django/cantusdb_project/main_app/templates/chant_search.html

```diff
@@ -209,8 +209,8 @@ <h3>Search Chants</h3>
                     {% for chant in chants %}
                         <tr>
                             <td class="text-wrap" style="text-align:left">
-                                {% if chant.source %}
-                                    <a href="{% url 'source-detail' chant.source.id %}" title="{{ chant.source.title }}">{{ chant.source.siglum }}</a>
+                                {% if chant.source__id %}
```

Comment:
I couldn't tell, but is this lookup optimized? I'm trying to remember if this is the case, but generally when doing double-underscore lookups, I like to make sure the target table (`source`, in this case) is added to a `select_related` so that it doesn't have to do the traversal.

(I can't actually test this with data, since I don't have a DB dump, so I'm just trying to go from memory)

Reply:
> is this lookup optimized?

I'm not sure what you mean.

I did some experiments with this, and couldn't get `select_related` or `prefetch_related` to play nicely with `values`:

If I specify `'source'` among the arguments to `.values`, I'll have access to the source's pk (i.e. the value in the 'source' column) but not to the source object itself. If I use `source` in the template after specifying it in the `values`, I'm worried future developers will think we're working with the source object rather than with its pk. Hence me using `source__id` in both places instead of `source`.

Since we're already fetching `source__title`, `source__siglum`, etc., postgres will already be joining the two tables under the hood, so I don't think fetching `source__id` will cause it to run any slower (i.e. as far as I can tell, it's just the difference between `SELECT chant.source AS 'source' FROM chant SOMEJOIN source ON chant.source = source.id` and `SELECT source.id AS 'source__id' FROM chant SOMEJOIN source ON chant.source = source.id`).

Does this address your question?

Reply:
I think I would need the data as well to test it. Let's say my spidey-senses were tingling, but I can't verify one way or the other. 

Note that `.values()` returns a [QuerySet of dictionaries](https://docs.djangoproject.com/en/4.1/ref/models/querysets/#values) which is probably why you couldn't get it to work properly. I'm not sure what problem you're trying to solve with using `.values()`, but maybe `.only()` [is useful too](https://docs.djangoproject.com/en/4.1/ref/models/querysets/#only)? This would return your results as objects, so it means you don't dead-end yourself into a list of dictionaries, and can keep spiralling down the ORM rabbit hole. :-D

Reply:
an update: I was not immediately able to make `select_related` work nicely with `only` - I wasn't able to use the ORM as I expected in the template, so values weren't rendering, and even with many things not rendered, the page didn't load any faster than when I used `.values`. When I asked it to count queries, DjDT counted just as many queries in both versions of the code.

Reply:
> DjDT counted just as many queries in both versions of the code.

out of curiosity, how many was it?

Reply:
11 queries both times! Our `base.html` template that all the other templates extend causes a handful of queries to run every time a page loads, so while I didn't look in detail, this is likely 10 queries for rendering the footer and 1 for the search results.